File: mobile-version/Navbar-shortened/app/src/main/java/com/example/clickanddrive/clients/services/AdminService.java
package com.example.clickanddrive.clients.services;

import com.example.clickanddrive.dtosample.requests.DriverRegistrationRequest;
import com.example.clickanddrive.dtosample.requests.NoteRequest;
import com.example.clickanddrive.dtosample.responses.AdminRideDetailsResponse;
import com.example.clickanddrive.dtosample.responses.AdminRideHistoryResponse;
import com.example.clickanddrive.dtosample.responses.AdminUserResponse;
import com.example.clickanddrive.dtosample.responses.DriverRegistrationResponse;
import com.example.clickanddrive.dtosample.responses.UserProfileResponse;

import java.util.List;

import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface AdminService {

    @POST("admin/drivers")
    Call<DriverRegistrationResponse> registerDriver(
            @Body DriverRegistrationRequest request,
            @Query("platform") String platform
    );

    @GET("admin/drivers/all")
    Call<List<UserProfileResponse>> getAllDrivers();

    @GET("admin/passengers/all")
    Call<List<UserProfileResponse>> getAllPassengers();

    @PUT("admin/users/{id}/block")
    Call<UserProfileResponse> blockUser(
            @Path("id") Long id,
            @Body NoteRequest request
    );

    @PUT("admin/users/{id}/unblock")
    Call<UserProfileResponse> unblockUser(@Path("id") Long id);

    @PUT("admin/users/{id}/note")
    Call<UserProfileResponse> leaveNote(
            @Path("id") Long id,
            @Body NoteRequest request
    );

    @GET("admin/ride-history")
    Call<List<AdminRideHistoryResponse>> getRideHistory(
            @Query("id") Long id,
            @Query("role") String role,
            @Query("sortBy") String sortBy
    );

    @GET("admin/users")
    Call<List<AdminUserResponse>> getAllUsers();

    @GET("admin/rides/{id}")
    Call<AdminRideDetailsResponse> getRideDetails(@Path("id") Long rideId);
}

File: frontend/ClickAndDrive/src/app/services/admin.service.ts
import { HttpClient } from "@angular/common/http";
import { NoteRequest, UserProfileInformation } from "./models/user-profile-information";
import { Observable } from "rxjs";
import { Injectable } from "@angular/core";
import { AdminRideHistory } from "./models/admin-ride-history";
import { AdminUser } from "./models/admin-user";
import { VehiclePrice } from "./models/vehicle-price";

@Injectable({
    providedIn: 'root'
})
export class AdminService {
    private apiUrl = 'http://localhost:8080/api/admin';

    constructor(private http: HttpClient) {}

    // GET all drivers
    getAllDrivers(): Observable<UserProfileInformation[]> {
        return this.http.get<UserProfileInformation[]>(`${this.apiUrl}/drivers/all`);
    }

    // GET all passengers
    getAllPassengers(): Observable<UserProfileInformation[]> {
        return this.http.get<UserProfileInformation[]>(`${this.apiUrl}/passengers/all`);
    }

    // Block a user
    blockUser(userId: number, reason: string): Observable<UserProfileInformation> {
        const body: NoteRequest = {message: reason};
        return this.http.put<UserProfileInformation>(`${this.apiUrl}/users/${userId}/block`, body);
    }

    // Unblock a user
    unblockUser(userId: number): Observable<UserProfileInformation> {
        return this.http.put<UserProfileInformation>(`${this.apiUrl}/users/${userId}/unblock`, {});
    }

    // Leave a note
    leaveNote(userId: number, note: string): Observable<UserProfileInformation> {
        const body: NoteRequest = {message: note}
        return this.http.put<UserProfileInformation>(`${this.apiUrl}/users/${userId}/note`, body);
    }

    getRideHistory(id: number, role: string, from?: string, to?: string, sortBy?: string) {

        let params: any = {
            id: id,
            role: role,
            sortBy: sortBy || 'date'
        };

        if (from) params.from = from;
        if (to) params.to = to;

        return this.http.get<AdminRideHistory[]>(
            'http://localhost:8080/api/admin/ride-history',
            { params }
        );
    }

    getAllUsers() {
        return this.http.get<AdminUser[]>('http://localhost:8080/api/admin/users');
    // Get current prices
    getPrices(): Observable<VehiclePrice> {
        return this.http.get<VehiclePrice>(`${this.apiUrl}/prices`);
    }

    // Update prices
    updatePrices(prices: VehiclePrice): Observable<any> {
        return this.http.put<any>(`${this.apiUrl}/prices`, prices);
    }
}